x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -273,12 +275,31 @@ public <A extends EsField> A readEsFieldWithCache() throws IOException {
                 return (A) esFieldFromCache(cacheId);
             }
         } else {
-            String className = readString();
+            String className = readCachedString();
             Writeable.Reader<? extends EsField> reader = EsField.getReader(className);
             return (A) reader.read(this);
         }
     }
 
+    /**
+     * Reads a cached string, serialized with {@link PlanStreamOutput#writeCachedString(String)}.
+     */
+
+    public String readCachedString() throws IOException {
+        if (getTransportVersion().before(TransportVersions.ESQL_CACHED_STRING_SERIALIZATION)) {
+            return readString();
+        }
+        int cacheId = Math.toIntExact(readZLong());
```

Comment:
Why use a long and not an int?

Reply:
I see - why not use a bool instead to signal when reading whether the follow-up read should be a string or a number?
You're saving one byte per cache hit however it seems to me you're losing that anyway through the int negation and conversion, despite zigzag?

Since the limit of entries is significantly under Integer.MAX you could use a bit (mask) instead of the bool however the savings might be minimal as it will impact the vint encoding.
Summoning @original-brownbear again :)

Reply:
I think a separate `bool` is not a good idea :) It's a full byte to encode the sign, while it's a single bit with zig zag. Performance while reading zLong, isn't amazing however. Maybe if performance is critical and since you're using an array here that you copy (which pretty much assumes that this is a very short list of strings), just use e.g. a short and get good performance for 32k values? That seems like a far faster option to me at the trivial price of an extra byte.
If you don't want a limit like that zLong is mathematically the most compact I think assuming an even distribution of values across positive and negative (note though that positive values are doubled in size, so using a short very quickly costs the same in terms of bytes).

Reply:
> Maybe if performance is critical and since you're using an array here that you copy (which pretty much assumes that this is a very short list of strings), just use e.g. a short and get good performance for 32k values?

32k could be too small for us in some cases.
Our main concern is compression rather than performance, but I'll try to measure the difference between the two and estimate the impact. 

> If you don't want a limit like that zLong is mathematically the most compact I think assuming an even distribution of values across positive and negative

The distribution will be very close to 50%-50% and it's hard for us to establish a maximum cache size (it would depend a lot on the complexity of the schemas of the queried indexes), so my guess is that ZLong is an acceptable solution.

Reply:
> I'll try to measure the difference between the two and estimate the impact.

I compared `main` with this PR with Nik's serialization tests, that are pretty heavy. There is no performance difference apparently.

Reply:
Given how per-request these streams are, do we expect that the caches on the writer and reader get out of sync, in the way the strings are added to them?
Otherwise could we not avoid the need to negate an ID and instead use a marker (like a compact 0?) to signal that a new value has been added on the writer and the reader should just do the same? I.e. count the existing cache size to obtain the new ID.
OR send the new ID as is and have the receiver add the following string to the cache on ID miss.